WIP: add asyncio

vsoch opened the pull request:
this will allow us to run two tasks using asyncio, and it is a start because it needs a few tweaks (I think) to make it better! Namely, we might want to handle a timeout > 0 in select via a flux Timeout(?) watcher, and we likely want to add proper tests and documentation (somewhere?) for the time being I have just put the example usage into the submit docstrings.

Thanks to @trws for two really fun pair programming sessions hacking on this!! asyncio is _so weird_!!! 

## Dinosaur TODO

- [x] I want to test the Timeout watcher to use with select to better honor the timeout passed (I have a prototype for this).
- [x] Where to write tests / what tests to write (this will be my first time, guidance appreciated!)
- [x] And documentation? Are flux docs all in an external repository from here?

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
oops did not lint. hehe. Let me see if I can fix that...

vsoch commented:
I don't think I can make pylint happy in all cases. For this one, I don't think I changed anything?
```console
Run ./scripts/pylint
************* Module flux.future
src/bindings/python/flux/future.py:56:0: R0902: Too many instance attributes (8/7) (too-many-instance-attributes)
```
and for this one, unfortunately it's the nature of asyncio - you have these cyclic imports to use the same loop, etc.

```console
************* Module flux.core.handle
src/bindings/python/flux/core/handle.py:1:0: R0401: Cyclic import (flux -> flux.core.handle -> flux.future -> flux.asyncio -> flux.asyncio.events) (cyclic-import)
src/bindings/python/flux/core/handle.py:1:0: R0401: Cyclic import (flux -> flux.core.handle -> flux.rpc -> flux.future -> flux.asyncio -> flux.asyncio.events) (cyclic-import)
```

trws commented:
If we decide it's being overzealous, we can override it for certain classes, which we might for those.

jameshcorbett reviewed (commented):
Glanced at it briefly and it looks really cool to me! Just one minor comment below.

Is this a WIP or are you looking for a comprehensive review?

jameshcorbett commented on src/bindings/python/flux/job/executor.py:
```diff
@@ -334,7 +334,10 @@ def result(self, *args, **kwargs):  # pylint: disable=arguments-differ
     def set_exception(self, exception):
-        """When setting an exception on the future, set the jobid if it hasn't
+        """
+        Set an exception.
```
Are the changes to this file necessary? Also, they don't seem related to the rest of the PR.

vsoch replied:
This was probably me mindlessly changing the comment style to one I like... sorry! I can change it back.

jameshcorbett replied:
It's just docstrings, but I would probably change it back, yeah. If you were going to make the change, it would probably go in another commit, but I'm not 100% certain about that.

vsoch commented:
@jameshcorbett this is an early draft to get more eyes and feedback! I'll update the description with the small list of TODOs (and those I might need guidance for).

jameshcorbett commented:
documentation is here: https://github.com/flux-framework/flux-core/tree/master/doc/python

Although whatever you put in your docstrings will go here: https://flux-framework.readthedocs.io/projects/flux-core/en/latest/python/complete_api.html , it's not the prettiest but it works OK

Tests are in `t/python/`, here is one example of many https://github.com/flux-framework/flux-core/blob/master/t/python/t0023-executor.py

vsoch commented:
oh the LGTM lemur is really cute, and I like the site too <3 
![image](https://user-images.githubusercontent.com/814322/192809878-a2658b7e-7e1a-47d0-aabc-1db4f5801427.png)

grondo commented:
> oh the LGTM lemur is really cute, and I like the site too <3

Too bad it is [deprecated](https://github.blog/2022-08-15-the-next-step-for-lgtm-com-github-code-scanning/)!

vsoch commented:
Nooooooo! 😭

vsoch commented:
okay this is likely not perfect - but I've taken a shot at all of the above!

 - the TimedWatcher is now used when timeout > 0 (and this seems to work for the asyncio.sleep)
 - I added a section to the docs for using it under job submission
 - I added a basic test that runs to ensure completion and times it

For docstrings, I restored them so the first line is in line with the parentheses, but I still left a newline between the top line and rest - I think the one line --> newline --> description is the suggested pep8 practice: https://peps.python.org/pep-0008/#documentation-strings.

https://github.com/flux-framework/flux-core/pull/4623/commits/bc022d3e2f7b1f5c933be99a7669636153f168d4

jameshcorbett commented:
> For docstrings, I restored them so the first line is in line with the parentheses, but I still left a newline between the top line and rest - I think the one line --> newline --> description is the suggested pep8 practice: https://peps.python.org/pep-0008/#documentation-strings.

That's true but actually the docstrings as written in the module are immediately overwritten by the following lines, e.g. `cancel.__doc__ = concurrent.futures.Future.cancel.__doc__`, which is why I wrote the docstrings to read more like developer notes than actual docstrings. But anyway, the changes you made are reasonable.

I'll start a full review now.

vsoch commented:
oh sorry totally missed that! I think it should still be useful for a developer (reading the code) but you're right - it probably doesn't matter too much.

jameshcorbett commented on src/bindings/python/flux/future.py:
```diff
@@ -185,6 +187,55 @@ def get(self):
+        import asyncio.exceptions  # pylint: disable=import-outside-toplevel
+        raise asyncio.exceptions.InvalidStateError('Result is not ready.')
+
+    __iter__ = __await__  # make compatible with 'yield from'.
```
What's the rationale for supporting `yield from`?

vsoch replied:
It's mostly because yield is also valid for asyncio (e.g., using `__iter__`) which I read here https://realpython.com/async-io-python/ and saw in several examples.

trws replied:
It's the old syntax for async generators, so it makes the futures compatible with more code across the pre/post asyncio boundary.

jameshcorbett commented on src/bindings/python/flux/future.py:
```diff
@@ -185,6 +187,55 @@ def get(self):
+        """
+        if self.is_ready():
+            return self.get()
+        import asyncio.exceptions  # pylint: disable=import-outside-toplevel
```
This import should go at the top level.

vsoch replied:
I put it here because I was worried if someone is using an older version of Python this would break their flux python bindings. Is this not the case / impossible to happen?

jameshcorbett replied:
Oh, I didn't realize it was a version thing. We support 3.6+ right now. It's better not to have version-specific code and it doesn't look like the exception you want exists in 3.6 (it comes in 3.7?) so I would just throw a generic `RuntimeError` or something.

vsoch replied:
This is moved to the top - non Python 3 users beware!

jameshcorbett replied:
Actually I think this will break 3.7 too. Like I said above I would actually remove this import entirely and instead just throw a generic RuntimeError or something.

vsoch replied:
oh gotcha], didn't see that. Will do.

grondo replied:
> non Python 3 users beware!

FYI - Flux requires Python >= 3.6

jameshcorbett commented on src/bindings/python/flux/future.py:
```diff
@@ -185,6 +187,55 @@ def get(self):
+            raise RuntimeError("await wasn't used with future")
+        return self.get()
+
+    def add_done_callback(self, func, *, context=None):
```
I don't think you should add methods to the existing `Future` class, because these methods might be mistakenly called by users who aren't using the asyncio interface, and that would be pretty confusing, especially given the kinds of errors that would result. I think it would be better to either wrap the `Future` in another class, or subclass it.

vsoch replied:
I thought the goal was to use the existing Flux future and just tweak slightly to work with asyncio - to be clear this absolutely won't work without these methods. Pinging @trws for comment because we made this decision together.

vsoch replied:
If we made another Future class (that adds the methods) it wouldn't plug nicely into submit_async (unless we did some weird fu)

jameshcorbett replied:
Yeah I take back my subclass idea, that doesn't make sense.

I guess what I was thinking is that instead of what you currently have:

```python

async def submit(fluxjob):
    uid = await flux.job.submit_async(handle, fluxjob)
    return uid
```

you could do something like


```python

class Wrapper:
    [stuff]

    def __init__(self, future):
       self.future = future

    def __await__(self)
        [stuff with self.future]


async def submit(fluxjob):
    loop = asyncio.get_running_loop()
    flux_handle = loop.flux_handle
    w = Wrapper(flux.job.submit_async(handle, fluxjob), loop)
    return await w
```

Somewhat related: when you need the loop, instead of trying to reach out to `flux.asyncio.loop`, you could just call `asyncio.get_running_loop()` which I think is more portable.

Also, if you make the loop's `flux.Flux()` handle a public attribute, you can enforce that `submit` and the loop` always have the same flux handle. 

I think this should actually work pretty nicely, but let me know if I'm missing something.

vsoch replied:
oh nice! Let me try that. Just to clarify - this should be like this right?
```diff
- flux_handle = loop.flux_handle
+  handle = loop.flux_handle
```
or I can just reference `loop.flux_handle` directly in the call.  I'll test the above now - I'm not sure how well it's going to wrok moving the await there but we'll find out!

vsoch replied:
Also where does the loop come in? it looks like you provided it as a second arg, but wrapper only takes 1 argument (the fluxjob).

jameshcorbett replied:
Yeah I just threw it together in a hurry, sorry for all the silly errors. I hadn't decided whether it would be better to pass the loop around all the time or instead to call `get_running_loop` whenever you needed it

vsoch replied:
okay it didn't like our wrapper:

> RuntimeError: Task got bad yield: <flux.job.submit.SubmitFuture object at 0x7fccaaa67b50>

vsoch replied:
Do you want me to push this somewhere for you to mess around with @jameshcorbett ?

vsoch replied:
You could probably just as easily start with the current version. I do want to restore what I had that was working (it was really hard to figure out).

vsoch replied:
okay - I updated to use asyncio.get_running_loop() - a huge improvement over the circular import! Nice!

jameshcorbett replied:
Yeah restore whatever you had, I will play around and perhaps propose a patch later, but I need to look at other stuff now.

vsoch replied:
okay sounds good! Thanks for your help today we made really good progress! :partying_face:

trws replied:
This would work, but I strongly prefer having our futures directly support the standard python futures interface.  Note that they generally are not asyncio specific.  This one currently is because of the invocation of loop, but we should be able to factor it into an use of either `.then` if the future is not ready or adding a `TimedWatcher` to the associated handle.  Actually, that's a question, @garlick what happens if you call `flux_future_then` or similar on a future that's already ready?

Either way, if we make it so that the future provides the "normal" python futures interface on top of flux features, in addition to our bespoke interface from before this stuff settled, I think we'd be better off overall *and* we get asyncio support without extra wrapper calls, just directly `await flux.job.submit()` or `loop.run_until_complete(asyncio.wait(loop.create_task(flux.job.submit())))`.

jameshcorbett replied:
> if we make it so that the future provides the "normal" python futures interface on top of flux features, in addition to our bespoke interface from before this stuff settled, I think we'd be better off overall and we get asyncio support without extra wrapper calls, just directly await flux.job.submit() or loop.run_until_complete(asyncio.wait(loop.create_task(flux.job.submit()))).

That makes a lot of sense actually, hadn't thought about it that way.

jameshcorbett commented on src/bindings/python/flux/future.py:
```diff
@@ -185,6 +187,55 @@ def get(self):
+        """
+        if not hasattr(self, "_loop"):  # pylint: disable=attribute-defined-outside-init
+            import flux.asyncio  # pylint: disable=import-outside-toplevel
+            self._loop = flux.asyncio.loop  # pylint: disable=attribute-defined-outside-init
```
Import should go at top level. Also, probably better to set the `_loop` attribute on the future when it is created rather than trying to look it up somehow and falling back to what might be the wrong loop.

vsoch replied:
This is actually the convention I saw in most implementations of asyncio. We have to do it at this level because otherwise we get a circular import. Also pinging @trws for comment!

jameshcorbett reviewed (commented):
OK, a lot of comments, most of them nits but some of them significant.

We're kind of picky about our commit history here so sorry about the git wrangling we'll ask you to do.

The commit messages should comply with Flux [RFC 1](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_1.html#patch-requirements), specifically:

-     commits should have a commit body. For example, adding a Problem: [Description] statement, then a description of how the problem is resolved by the commit.
-     commits should be in imperative, e.g. "add expected outputs" not "adding expected outputs"

vsoch commented on doc/python/job_submission.rst:
```diff
@@ -94,9 +94,10 @@ restricted to instance owners.
+There are two primary asynchronous interfaces to job manipulations, and a third
+that uses Python asyncio to submit jobs. The first is an event-loop interface,
+which is closer to the native C interface, and consists of functions like 
+``flux.job.submit_async`` and ``flux.job.result_async`` (note the
```
Okay I'll try to do a general overview - it's pretty complex so I'll give a high level understanding. I'm not sure I fully understand the pros/cons of each yet, beyond saying this third way is using Python's native implementation (interacting with the Flux futures).

vsoch commented on doc/python/job_submission.rst:
```diff
@@ -197,3 +198,43 @@ as futures complete.
+with a Flux future (without asyncio).Note that we also use a common event loop 
+and handle that ``flux.asyncio`` defines. 
+You are, however, free to use the ``FluxEventLoop`` class to define your own, using
+your own handle.
```
yes, I'll clarify that!

vsoch commented on etc/docker/ubuntu/README.md:
````diff
@@ -11,6 +11,12 @@ Build the main `flux-ubuntu` image:
+
+```bash
+$ docker run -it -v $PWD:/code ghcr.io/flux-framework/flux-ubuntu
+```
````
This is just showing exactly how you can develop the Python stuffs (as I've done in this PR) using the ubuntu container. I think it makes sense to add here because someone would need to know this to develop (at least in a container - I've actually never used/developed flux outside of one!)

jameshcorbett replied:
Yeah that makes sense, I still think it should go in a separate commit. I'm not an authority on the interpretation of RFC 1 though, I'm just interpreting what I've learned from the feedback I got on my own PRs. Will let @grondo lay down the law 😉

vsoch commented on src/bindings/python/flux/asyncio/__init__.py:
```diff
@@ -0,0 +1 @@
+from .events import FluxEventLoop, loop, submit
```
I think my preference is generally for relative imports - it's not a strong preference, but they look more succinct, are more portable (if I decide to move as a developer) and I don't know if this is a real concern or just an anxiety I have - but without relative, someone could add a different "flux" (or module name) to the PYTHONPATH to be found before wherever the main flux is, and then import from there instead of the expected location. Perhaps the relative import helps prevent that?

vsoch replied:
This imports conversation can be followed up in the thread below (we currently have two).

vsoch commented on src/bindings/python/flux/job/submit.py:
```diff
@@ -25,6 +25,10 @@ def get_id(self):
+    def get(self):
+        """Return the job represented by this future."""
+        super().get()
+        return submit_get_id(self)
```
Can you explain why? The flux async won't work without it - it was completely necessary / required for it.

jameshcorbett replied:
Even if it was necessary, it could go in a prior commit, right? But I don't see how it is necessary. Anywhere you call the new `f.get()` implementation you could have called `submit_get_id(f)` instead.

vsoch replied:
Yeah we could have - I wrote this with @trws and I think we liked adding the get() function. But if you absolutely hate it I guess we could remove it and call those two things separately? It would still be kind of wonky because the submit_get_id needs a self, so I like this design better.

jameshcorbett replied:
I like the design too, I was just asking you to put it in a separate commit. It sounds like you want to do the git history rewriting later though, which is OK.

grondo replied:
Isn't the `super().get()` implied by `submit_get_id()`? Someone reading the code here could get confused, so at least a comment would be helpful.

 BTW, I think the Python here was just following the C API, which needs a separate `flux_*_get_something()` to pull correct types out of a `flux_future_t` since there's no subclassing in C. I think a lot of other Flux Python APIs probably follow this same idiom. If we override `.get()` in the `SubmitFuture` class, perhaps we should do it the other subclasses of Future as well?

grondo replied:
(not a strong opinion here, just bringing up the question)

vsoch replied:
Could you tell me exactly which classes you want it added to?

trws replied:
We added it because we need a way to implement `.result()`.  I was actually surprised when I realized we didn't have a unified `get` method, that's what `get` was supposed to do, but that's why this is here.  Without it the `.result()` method would have to become some wretched dystopian horror of `isinstanceof` crossed with custom method names.  I was hoping we could avoid "result as implemented by Heironymus Bosch," so *all* futures must implement a common getter method, so we added `.get`.

As to the specific issue, technically yes it could be moved to a separate commit, but it was part of the implementation of result, so I'm not sure it's worth it unless the goal is to add a note to the commit message.

Now I really want to try "implementation of result getter method by Heironymus Bosch" in Stable Diffusion...

vsoch commented on src/bindings/python/flux/asyncio/events.py:
```diff
@@ -0,0 +1,66 @@
+
+# Make sure submit async and we see job go through
+# Ensure the job and loop use the same handle
+handle = flux.Flux()  # pylint: disable=invalid-name
```
haha ok I can DO THAT :laughing:  Python linting is so strange! I guess it's a good visual indicator of a global variable though!

vsoch commented on src/bindings/python/flux/asyncio/events.py:
```diff
@@ -0,0 +1,67 @@
+
+
+# This loop needs to be accessible from all places!
+loop = FluxEventLoop(handle)  # pylint: disable=invalid-name
```
I'm not sure how to go about it, because we need to have a shared loop that is importable from the flux.job code too. It would work fine if the user redefines it, but if we don't have it there, we can't do the second important. It's a bit wonky how asyncio needs to do that. Do you have an idea for how else to go about it?

jameshcorbett replied:
You mean because Futures reach out for the `loop` attribute? I think I have another comment about how that could be avoided.

vsoch replied:
yep - we need to be using the same loop instance across the different files, and we need to avoid circular import. What is your idea?

jameshcorbett replied:
https://github.com/flux-framework/flux-core/pull/4623#discussion_r982855682

Resolving this issue (to continue it in that thread, maybe)

vsoch replied:
Okay this is fixed to use `asyncio.get_running_loop()`

vsoch commented on src/bindings/python/flux/asyncio/events.py:
```diff
@@ -0,0 +1,66 @@
+
+    Example usage
+    =============
+    # ensure you've started flux
```
In the container it would be `flux start --test-size=4` - if you don't do that you get a bunch of errors.

jameshcorbett replied:
Oh, yeah this is a bit confusing. Hmmmm. I think maybe it should read something like "ensure the script is running in an active flux instance"

vsoch replied:
ok!

vsoch commented on src/bindings/python/flux/asyncio/events.py:
```diff
@@ -0,0 +1,66 @@
+
+    def __init__(self, flux_handle=None):
+        if not flux_handle:
+            flux_handle = handle
```
I originally did that so submit() and the loop would have the same handle - is that not important? E.g., flux.Flux() wherever you call it / multiple times returns the same handle?

jameshcorbett replied:
No, you would get a different handle every time. I guess I was thinking---if they strictly need to be the same handle, then why is `flux_handle` accepted as an optional argument?

vsoch replied:
Update for here - it stops working when you don't use a common handle for the submit and the loop.

vsoch replied:
That's an "advanced" use case - technically someone could create some custom handle and then provide to their submit and loop. My expectation (why I say advanced) is that most people won't. But I think it could be an edge case, and we can easily support it with this method.

trws replied:
If we want multiple handles to work, we'll need to re-design select a bit, and... well, everything else.  There needs to be a handle for the Selector instance that it will always add its watchers to.  If we want to allow flux futures to be incorporated that use different handles, we'll need to detect that case when they get added and never actually block in reactor_run but loop over them instead.  Unless, @grondo or @jameshcorbett do we have a way to add a handle to another reactor?

grondo replied:
Yes, in the C API we have [`flux_handle_watcher_create(3)`](https://flux-framework.readthedocs.io/projects/flux-core/en/latest/man3/flux_handle_watcher_create.html)

trws replied:
Excellent, then yeah we could actually support that reasonably, we just have to figure out how to make sure that we get that done when a future wrapped in a task gets added to the event loop. Maybe we add a check into the __await__ method to see if the loop uses the same handle and if not add the future’s handle to it across the yield?

vsoch replied:
okay so conceptually - the event loop has one internal handle, and when a new task is added, when we hit the future.__await__ we say:

1. is this the same flux_handle the loop has? If yes, carry on!
2. if no, we need to switch the handle to be the loop's (e.g., future.flux_handle = loop.handle) OR do we need to somehow steal the watchers for that handle and add to the one loop handle, or something else?

vsoch replied:
okay - so I played with this for a bit, and came up with https://github.com/researchapps/flux-core/compare/test/asyncio...researchapps:flux-core:test/asyncio-updates?expand=1. The issue that we run into is that even though we can check for handle equality and then create the new HandleWatcher, there is still no connection back to the file descriptors and SelectKey that asyncio has registered, so the resulting outcome is that the loop runs forever because it can never know that it's ready. I chat with @trws and we decided that trying to get this working would be going down a rabbit hole of potentially huge complexity, and for now we should add a check to ensure that the handles are the same, and throw an error if not. And we can come back to this in a round 2 if we absolutely find a case where two different flux handles must be supplied.

jameshcorbett commented on src/bindings/python/flux/asyncio/events.py:
```diff
@@ -0,0 +1,66 @@
+import flux
+import flux.core.watchers
+
+from .selector import FluxSelector
```
I doubt anyone else has strong feelings either. But all of the other imports in the package are absolute if that makes a difference. And the general community feeling is [in favor of absolute (with a few exceptions)](https://stackoverflow.com/questions/4209641/absolute-vs-explicit-relative-import-of-python-module). And if anyone messes with `sys.path` to add another 'flux' module there will be all kinds of things breaking that absolute vs. relative imports would not affect at all.

Anyway, those are my thoughts, maybe @grondo has some? But if you want to keep the relative imports, feel free to mark these conversations as resolved and keep them, it's not worth spending much time on.

vsoch replied:
All hail @grondo !! We shall follow your guidance!

grondo replied:
I'm not really a Python expert, so I'd defer to @jameshcorbett here.

vsoch replied:
It's mostly a preference I think. I've been writing Python only about 13 years, so I'm happy to go with @jameshcorbett preference.

jameshcorbett replied:
I don't know that I've been alive for 13 years so my preference is that you do what you want!

jameshcorbett replied:
(In other words, keep the relative imports pls)

vsoch replied:
lol! Look out, here comes the old fart Dinosaur! :dash:

vsoch commented on t/python/t0008-asyncio.py:
```diff
@@ -0,0 +1,61 @@
+    Tests for Flux Asyncio.
+    """
+
+    def test_as_completed(self):
```
yeah I should have known that :)

I do think it makes sense to keep the sleep - the asyncio.sleep is using the timeout functionality, and if our implementation is wrong (or breaks) it could return before that is complete (which we'd want to know).

vsoch commented on src/bindings/python/flux/asyncio/selector.py:
```diff
@@ -0,0 +1,153 @@
+    """
+    This is called by the watcher, and stops the reactor add adds ready jobs.
+    """
+    _ = watcher
```
This was me trying to make the linter happy - it doesn't like unused variables in functions, but the callback has an expected signature. I can't make both args `_` (another lint error) so instead I define one and then "send it into the abyss" hehe

vsoch replied:
okay I'll try that! I do think it's useful to preserve some information about what those arguments are, for a future developer reader.

vsoch commented on src/bindings/python/flux/asyncio/selector.py:
```diff
@@ -0,0 +1,153 @@
+    reactor = handle.get_reactor()
+    handle.reactor_stop(reactor)
+
+    # TimedWatcher won't know a file descriptor, just needs to stop the reactor
```
haha, I literally kept changing this back and forth! Hopefully I'll remember the right one soon ;)

vsoch commented on src/bindings/python/flux/asyncio/selector.py:
```diff
@@ -0,0 +1,153 @@
+
+                # 0 == "run until I tell you to stop"
+                # 1 == "no wait"
+                # 2 == "once"
```
oh nice! Tom and I were looking for those (and didn't find them!)

vsoch replied:
I don't see one for the value of 0 - is the name / prefix different? I'll leave as  0 for now.

jameshcorbett replied:
For 0 just don't pass any argument I'd say (0 is the default).

vsoch replied:
ah boum!

vsoch replied:
To complete this thread - I learned that I'll get an error from the underlying flux library about the wrong number of arguments without having two, and without having them as positional arguments. If we wanted to do this we'd need another level of wrapper.

jameshcorbett replied:
That sounds good to me, but also sounds like a bug! If you could make a reproducer and post an issue I'd appreciate it!

vsoch commented on src/bindings/python/flux/asyncio/selector.py:
```diff
@@ -0,0 +1,153 @@
+        `events` is a bitwise mask of EVENT_READ|EVENT_WRITE
+        """
+        # Store the watcher as data
+        reactor = self.handle.get_reactor()
```
This one needs the flags for the second argument, and this is also how I see it used elsewhere, so I think it would be OK to leave like this to be consistent.

jameshcorbett replied:
You could pass the flags as a keyword argument `(flags=...)` but OK.

vsoch replied:
okay, happy to do that too!

vsoch replied:
Actually I take it back - I had tried this and it doesn't work. If you provide flags, it gets angry that the argument isn't correct. If I look at all other use cases in the code, two args are provided. If I stray from that:

```python
WrongNumArguments: 
The wrong number of arguments has been passed to wrapped C function:
Expected 2 arguments, received 0
Name: flux_reactor_run
C signature: int(*)(struct flux_reactor *, int)
Arguments: []
Handle type: <ctype 'struct flux_handle_struct *'>
```
So I probably put the reactor and flags there as I did because that's how I got it to work!

vsoch replied:
Going to resolve this - unresolve if there is another way you want to try!

vsoch commented on src/bindings/python/flux/asyncio/selector.py:
```diff
@@ -0,0 +1,153 @@
+    return event
+
+
+class FluxSelector(selectors._BaseSelectorImpl):  # pylint: disable=protected-access
```
yes! So I wanted the most bare selector that would be possible - without losing some useful functionality. If you look at the source code https://cs.github.com/python/cpython/blob/68c46ae68b6e0c36a12e37285fff9ce0782ed01e/Lib/selectors.py?q=_BaseSelectorImpl#L206 if we go one level, we lose nice courtesy functions in _BaseSelectorImpl that do the file descriptor lookup for us. We'd have to reimplement that, either copying it or something simpler (that maybe doesn't work as well). That also gives us `self.modify` which I haven't encountered hitting yet, but if we do hit it for some case that's what we'd want it to do (it's not implemented in the base class one level up). Finally, the next level down (SelectSelector and other variants) are tuned for specific kinds of events - we overwrite most of those classes so we could technically use it (because it's "public") but really we are using the _BaseSelectorImpl, which is the best match/level for what we need.

The distinction of it being "private" does indicate to external developers that it's not intended for this kind of use, however, so if you have strong feelings about that I can switch it to be the one level up public class, understanding that we aren't using most of it anyway.

jameshcorbett replied:
It sounds reasonable, but... why not the `DefaultSelector`? If the only trade-off is that we're overwriting some of its functionality. There's nothing wrong with some overwriting.

vsoch replied:
DefaultScheduler is provided on the bottom - and you don't really know which one you are getting! Let's ask @trws about this one, I don't have a good sense of it it matters. I wanted to choose the most correct one for the job, which is exactly the `_BaseSelectorImpl` and the others are doing tweaks that we don't really need (and could lead to a bug, it's hard to say).

trws replied:
I specifically requested to use _BaseSelectorImpl, at least while we're developing on this, because if we use DefaultSelector and miss something then (at least at first) things appeared to work when they were using the wrong loop, and yet _BaseSelectorImpl has some useful helpers in it.  We should probably switch to BaseSelector at some point.

vsoch commented:
Thanks for the review! For the commit / messages, I think it's easier to do a ton of review (force pushing commits is ok) and then doing commit --amend for the final message. It's hard to say how things will change in the review process, and it's easier to do all this at the end than every time!

vsoch commented:
okay - just tested changed to having different handles - it borked the entire thing (the loop never ended). We absolutely need the same handle, which is what my original comment noted.

vsoch commented on src/bindings/python/flux/future.py:
```diff
@@ -99,7 +99,9 @@ def __init__(self, future_handle, prefixes=None, pimpl_t=None):
     def stop(self):
-        """Stop a future from calling the user callback.
+        """
+        Stop a future from calling the user callback.
```
No you are totally right - adding the newline is wrong! Now I have like a gazillion libraries to fix, haha.

grondo replied:
A good rule is to follow [RFC 1 Patch Requirements](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_1.html#patch-requirements), specifically this bullet:

> A patch SHOULD be a minimal and accurate answer to exactly one identified and agreed problem.

A change that trivially modifies a comment block does not solve an identified problem, unless the comment is being fixed or updated for correctness, or the formatting changed to match some new requirement. In both of these cases, the comment should be updated it its own commit (a patch should be minimal and answer one problem), and should come with a commit message that states the problem being solved.

trws commented:
I was originally thinking something like that, but the function @grondo reminded me of above lets us get a file descriptor that represents “ready” for an entire handle when it shows ready. So we can add the entire handle as a watcher on the loop’s handle with a callback that runs a once-through on its reactor, I think. If that works then we won’t have to worry about it in the loop at all, just ensure that the futures register themselves correctly before yielding. That would mean even communicating with separate flux instances in the same loop should work.


---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>

On September 28, 2022 at 9:40:43 PM EDT, Vanessasaurus ***@***.***> wrote:

@vsoch commented on this pull request.

________________________________

In src/bindings/python/flux/asyncio/events.py<https://urldefense.us/v3/__https://github.com/flux-framework/flux-core/pull/4623*discussion_r982989537__;Iw!!G2kpM7uM-TzIFchu!npeee8wIFjVeP2fBv-n9Mf-tw-pICvsTe7t-aSVzjJ38GC-s4Rjw_htoH2QKh8NUWw$>:

> +    asyncio.set_event_loop(loop)
+    results = loop.run_until_complete(asyncio.gather(*tasks))
+    # [JobID(456004999315456), JobID(456004999315457)]
+    """
+    uid = await flux.job.submit_async(handle, fluxjob)
+    return uid
+
+
+class FluxEventLoop(asyncio.SelectorEventLoop):
+    """
+    An asyncio loop that handles running Flux.
+    """
+
+    def __init__(self, flux_handle=None):
+        if not flux_handle:
+            flux_handle = handle


okay so conceptually - the event loop has one internal handle, and when a new task is added, when we hit the future.await we say:

  1.  is this the same flux_handle the loop has? If yes, carry on!
  2.  if no, we need to switch the handle to be the loop's (e.g., future.flux_handle = loop.handle) OR do we need to somehow steal the watchers for that handle and add to the one loop handle, or something else?

—
Reply to this email directly, view it on GitHub<https://urldefense.us/v3/__https://github.com/flux-framework/flux-core/pull/4623*discussion_r982989537__;Iw!!G2kpM7uM-TzIFchu!npeee8wIFjVeP2fBv-n9Mf-tw-pICvsTe7t-aSVzjJ38GC-s4Rjw_htoH2QKh8NUWw$>, or unsubscribe<https://urldefense.us/v3/__https://github.com/notifications/unsubscribe-auth/AAFBFNNRIFKZXUFZP6ZH2KTWATXRLANCNFSM6AAAAAAQXGIWJI__;!!G2kpM7uM-TzIFchu!npeee8wIFjVeP2fBv-n9Mf-tw-pICvsTe7t-aSVzjJ38GC-s4Rjw_htoH2S9hPhTYw$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

vsoch commented:
okay so I understand (and before trying to implement) you are saying:

Forget about adding the handle when the loop is init - we will get handles when the events come in. So the init becomes:

```python
class FluxEventLoop(asyncio.SelectorEventLoop):
    """An asyncio loop that handles running Flux.
    """

    def __init__(self):
        selector = FluxSelector()

        # Reverse reference is needed for watchers
        selector.loop = self
        super().__init__(selector)

```
No flux handle there!  And then if we need to get access to the underlying job (because in the register function we can only see the file descriptor) we could look at loop.create_task:

https://github.com/python/cpython/blob/6281affee6423296893b509cd78dc563ca58b196/Lib/asyncio/tasks.py#L371-L384

which is here:

https://github.com/python/cpython/blob/487135a396a504482ae3a7e9c7f80a44627a5a3f/Lib/asyncio/base_events.py#L429-L448

so are you saying we could get the file descriptor possibly there (while we still have access to the job) and then call register ourselves to create the watcher? And then the watcher would be associated with "whatever flux handle they provided" and we wouldn't have to be so ornery about them providing the same one? And then if register is still called for file descriptors that aren't called by us (and we'd need some flag) do we ignore them or do a different action?

trws commented:
I misspoke, and made a bit of a mistake last night (what I get for trying to program into an email client on an iphone).  I think we do need a handle that's part of the Selector, the EventLoop probably doesn't have to have it but the selector does.  Then in `__await__` on the future I think we can do something like this (pseudopython, there will be mistakes, also we need to implement the python wrapper HandleWatcher since we don't have one):

```python
loophandle = self._loop.selector.handle
def process_ready_loop(handle):
  handle.reactor.run(NOWAIT)
if self.handle != loophandle:
  self._watchers.append(flux.HandleWatcher(loophandle, self.handle, process_ready_loop))
yield self
if self.handle != loophandle:
  self.remove_handle_from_watchers()
```

We may want to reference count the handle watchers, in case we get multiple, but worst case this would have some overhead watching the same reactor more than once.

vsoch commented:
ohhh I think I get it! So we want to (when we hit the future) check it's handle against the selector, if it's the same we're good (we are already watching) but if not, we need to update to use the watcher that it's linked to (and throw away the other one). I'll munch on this more a little later today - currently in Go land with all the cute gophers and very happy :)

trws commented:
Almost yeah, but we actually *add* the new one to the current one, and keep watching the current one.  That way they both make progress.  Does that make sense?

vsoch commented:
Yep. Question though - the watcher that we create (that isn't associated with the correct handle) what exactly is it watching?

trws commented:
There's a special watcher *to watch other handles*, that's the HandleWatcher, so it lets us add a second handle to the current one.  Doing that with a callback that will process its events should let us add the "foreign handle" or what have you and watch it until the future being awaited is fulfilled, then we remove that watcher from the current loop.  There may be more efficient ways to do this, but I'm relatively sure it will work correctly.

vsoch commented:
Just a comment (criticism) coming from developing many Python projects with large communities over time! Linting is great - but there is a balance between "ensure code quality" and "making it so awful and annoying to contribute I never want to do it again." The ideal middle ground is to provide tooling that can fix issues. But for example, having to put all these pylint ignore lines? Or getting griped about too many class attributes (for a class I seemingly didn't change?) I think this is a significant issue because it makes it a really high bar to contribute, and people just won't want to.

vsoch commented:
Specifically:

### black

People will tend to create virtual environments or similar to develop, and not have the "right" version of black, then they discover through the CI the black linter is wrong, and are confused for a while until they dig in and find that it's pinned somewhere. Why is it pinned? 

### pylint

And then for the pylint stuff, you start seeing this:

```python
# There isn't anything wrong with this
if future_handle.handle != loop_handle.handle:    
    raise RuntimeError("Found different future and loop handle - ensure the same.")
```
but actually, there is! It's a few characters too long. So then you see (imho, uglier) code like this:

```python
if future_handle.handle != loop_handle.handle:
    msg = "Found different future and loop handle - ensure the same."
    raise RuntimeError(msg)
```

and then it passes. I know it supposedly makes it easier to have standards, but if it means the developer is pushing 8-10 times to fix things when otherwise it could just be a handful, that's a negative developer experience. These nits only make sense if you can automate commands to fix them automatically, imho. It would be a compromise to have something like LGTM _flag_ potential issues, and then have the maintainers (that are already doing a comprehensive review) decide which are actually meaningful to impact the quality of the project, and which are probably just noise that won't make a difference in the long run.

Perhaps there is a hard coded decision here to "use pep8 and no other way!" but I wanted to bring this up for discussion because (as a contributor) this is really a bottleneck.

vsoch commented:
So my suggestion (so I'm not just a complainysaurus!) is that we have a meeting (or async discussion) to:

1. make a list of the linting we want, rank by importance
2. decide where we can relax a bit - or have some middle ground (e.g., warnings that a maintainer checks)
3. for absolutely necessary, figure out how to provide tooling / automation to make fixing easier.

grondo commented:
> but there is a balance between "ensure code quality" and "making it so awful and annoying to contribute I never want to do it again." The ideal middle ground is to provide tooling that can fix issues. 

I agree. Let's fix some of these painful and honestly difficult to defend pylint requirements in the `.pylintrc`, especially if you find it necessary to add a bunch of pylint ignore comments. Honestly, I'm not a big Python developer and just assumed most projects used `pylint` and dealt with the fallout.

> People will tend to create virtual environments or similar to develop, and not have the "right" version of black, then they discover through the CI the black linter is wrong, and are confused for a while until they dig in and find that it's pinned somewhere. Why is it pinned?

I am not a big fan of dictatorial formatting either, but someone else chose to add `black`. We pin the version because otherwise you can get failures in CI for PRs _that don't even touch Python_ because the version of black was updated and they changed their minds. This would be a lot more painful for developers than specifying a given version.

vsoch commented:
> I am not a big fan of dictatorial formatting either, but someone else chose to add black. We pin the version because otherwise you can get failures in CI for PRs that don't even touch Python because the version of black was updated and they changed their minds. This would be a lot more painful for developers than specifying a given version.

I actually think black is ok - but the releases are done fairly often. I've experimented with both approaches - pinning and not pinning, and I have (maybe against intuition) found that not pinning it leads to less buggy outcomes mainly because people are typically installing the latest into their virtual environment. For these cases when black is updated, I think automation could easily fix this case - e.g., have a nightly black check that, given changes, updates the repository with them.

Some approaches that I've seen (and like):

- providing a pre-commit recipe, e.g., river does all the checks for you on commit, so a lot of the fixing is automated, and for things I need to fix I see them right away instead of waiting for the CI: https://github.com/online-ml/river/blob/main/.pre-commit-config.yaml
- spack style is an example of integrating into spack, but I think that's probably overkill because I don't want the linting to be a part of my software.

Would others be interested in trying out pre-commit hooks (or another automated way for developers to perform these checks?)

tgamblin commented:
FWIW and maybe y'all already fixed this, but shouldn't `black` fix this:

```python
if future_handle.handle != loop_handle.handle:    
    raise RuntimeError("Found different future and loop handle - ensure the same.")
```
to this:

```python
if future_handle.handle != loop_handle.handle:    
    raise RuntimeError(
        "Found different future and loop handle - ensure the same."
    )
```

It seems like either `black` isn't getting run before `pylint`, or the two aren't set to use the same line length.

grondo commented:
I've never seen pylint complain about long lines :shrug:

vsoch commented:
I was anxious about doing a rebase, so made a new branch (and will open the updated PR for it).
